src/Symfony/Component/ObjectMapper/ObjectMapper.php

```diff
@@ -0,0 +1,226 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\ObjectMapper;
+
+use Psr\Container\ContainerInterface;
+use Symfony\Component\ObjectMapper\Attributes\Map;
+use Symfony\Component\ObjectMapper\Exception\RuntimeException;
+use Symfony\Component\PropertyAccess\PropertyAccessorInterface;
+
+/**
+ * Object to object mapper.
+ *
+ * @implements ObjectMapperInterface<T>
+ *
+ * @template T of object
+ *
+ * @experimental
+ *
+ * @author Antoine Bluchet <soyuka@gmail.com>
+ */
+final class ObjectMapper implements ObjectMapperInterface
+{
+    public function __construct(
+        private readonly MapperMetadataFactoryInterface $metadataFactory = new ReflectionMapperMetadataFactory(),
+        private readonly ?PropertyAccessorInterface $propertyAccessor = null,
+        private readonly ?ContainerInterface $serviceLocator = null
+    ) {
+    }
+
+    public function map(object $object, object|string $to = null): object
+    {
+        static $objectMap = null;
+        $objectMapInitialized = false;
+
+        if (null === $objectMap) {
+            $objectMap = new \SplObjectStorage();
+            $objectMapInitialized = true;
+        }
+
+        $refl = new \ReflectionClass($object);
+
+        $metadata = $this->metadataFactory->create($object);
+        $map = $this->getMapTo($metadata, null, $object);
+        $to ??= $map?->to;
+
+        if (!(\is_string($to) && class_exists($to)) && !\is_object($to)) {
+            throw new RuntimeException(sprintf('Attribute of type "%s" expected on "%s.', Map::class, $refl->getName()));
```

Comment:
Can this be a specific exception ? like `InvalidMapTargetException`

Reply:
Also from what i understand, transform can return a completely different object and this will still work I'm not sure if it should allow that as the API will not be respected in this case (it will not return a `T`) maybe we should add an additional check here to ensure that object transformed is a `T` (or extends it)

Reply:
but transform doesn't necessarily return an object, I like that it's flexible. Resolving this as I added a specific exception but feel free to reopen the discussion on that